toolchain/check/member_access.cpp

```diff
@@ -93,12 +90,111 @@ static auto IsInstanceMethod(const SemIR::File& sem_ir,
+                                 SemIR::NameScopeId name_scope_id) -> bool {
+  auto inst_id = context.name_scopes().GetInstIdIfValid(name_scope_id);
+  if (!inst_id.is_valid()) {
+    return false;
+  }
+
+  auto inst = context.insts().Get(inst_id);
+  if (inst.Is<SemIR::InterfaceDecl>()) {
+    // Don't perform impl lookup if an associated entity is named as a member of
+    // a facet type.
+    return false;
+  }
+  if (inst.Is<SemIR::Namespace>()) {
+    // Don't perform impl lookup if an associated entity is named as a namespace
+    // member.
+    // TODO: This case is not yet listed in the design.
+    return false;
+  }
+  // Any other kind of scope is assumed to be a type that implements the
+  // interface containing the associated entity, and impl lookup is performed.
+  return true;
+}
+
+// Given a type and an interface, searches for an impl that describes how that
+// type implements that interface, and returns the corresponding witness.
+// Returns an invalid InstId if no matching impl is found.
+static auto LookupInterfaceWitness(Context& context,
+                                   SemIR::ConstantId type_const_id,
+                                   SemIR::InterfaceId interface_id)
+    -> SemIR::InstId {
+  // TODO: Add a better impl lookup system. At the very least, we should only be
+  // considering impls that are for the same interface we're querying. We can
+  // also skip impls that mention any types that aren't part of our impl query.
+  for (const auto& impl : context.impls().array_ref()) {
+    if (context.types().GetInstId(impl.self_id) != type_const_id.inst_id()) {
+      continue;
+    }
+    auto interface_type =
+        context.types().TryGetAs<SemIR::InterfaceType>(impl.constraint_id);
+    if (!interface_type) {
+      // TODO: An impl of a constraint type should be treated as implementing
+      // the constraint's interfaces.
+      continue;
+    }
+    if (interface_type->interface_id != interface_id) {
+      continue;
+    }
+    // TODO: Diagnose if the impl isn't defined yet?
+    return impl.witness_id;
+  }
+  return SemIR::InstId::Invalid;
+}
+
+// Performs impl lookup for a member name expression. This finds the relevant
+// impl witness and extracts the corresponding impl member.
+static auto PerformImplLookup(Context& context, SemIR::ConstantId type_const_id,
+                              SemIR::AssociatedEntityType assoc_type,
+                              SemIR::InstId member_id) -> SemIR::InstId {
+  auto witness_id =
+      LookupInterfaceWitness(context, type_const_id, assoc_type.interface_id);
+  if (!witness_id.is_valid()) {
+    CARBON_DIAGNOSTIC(MissingImplInMemberAccess, Error,
+                      "Cannot access member of interface {0} in type {1} "
+                      "that does not implement that interface.",
+                      SemIR::NameId, std::string);
+    context.emitter().Emit(
+        context.insts().GetNodeId(member_id), MissingImplInMemberAccess,
```

Comment:
```suggestion
        member_id, MissingImplInMemberAccess,
```

Just a reminder, but GetNodeId should rarely be required. Probably I should also change TODO to use a SemIRLocation.